witte_jacquelyn_ndvi.py

```diff
@@ -0,0 +1,898 @@
+- Calculate the NDVI
+- Calculate the mean NDVI
+- Turn it into a dataFrame
+- Plot it
+
+The idea is to create many functions to perform most of the outlined steps above. When looping through all the data directories to calculate a single mean NDVI value and grab it's metadata, the sequence of code will be a series of function calls only.
+
+Note, that I am choosing to read all the bands so there is the option to look at RGB or CIR in the future. 
+
+Function 1 - Retrieve tif files
+
+    bands_tif = sorted(os.path.join(landsat_path,'*band[2-5]*.tif'))
+    qa_tif = os.path.join(landsat_path,'*qa*.tif')
+
+Function 2 - Key metadata: date and site
+
+    site = filepath.split('/')[2]
+    date = datetime.strptime(filepath.split('T')[0][-10:-2], "%Y%m%d")
+
+Function 3 - Consolidated Landsat image, cropped and cloud-free
+
+    # For loop over tif files and append to a list.
+    all_bands = []
+    for i, aband in enumerate(files):
+        all_bands.append(rxr.open_rasterio(aband, masked=True).squeeze())
+        # Assign a band number to the new xarray object
+        all_bands[i]["band"] = i+1
+        
+    # Turn list of bands into a single xarray object
+    - Create xarray.DataArray with xr.concat()
+    
+    # Crop the data to the shapefile
+    - Use dataArray.rio.clip(crop_boundary.geometry) method
+    
+    # IMPORTANT: Clean the dataArray - remove 0's and high values
+    # I went with a less elegant but explicit and clear code that I can understand. 
+    data_nozeros_xr = data_xr_crop.where(data_xr_crop > 0, np.nan)
+    data_nozeros_xr = data_nozeros_xr.where(data_nozeros_xr < 10000, np.nan)
+
+Function 4 - Apply cloud mask
+
+Reference: https://github.com/earthlab-education/ea-python-course-notebooks/
+                       blob/main/2022/completed-demos/05-l1-landsat-cloud-masks.ipynb
+
+    # Read the QA file
+    qa_xr = rxr.open_rasterio(qa_file).squeeze()
+    
+    # Create cloud mask using earthpy mask package for Landsat imagery
+    high_cloud_confidence = (
+        em.pixel_flags["pixel_qa"]["L8"]["High Cloud Confidence"])
+    cloud = em.pixel_flags["pixel_qa"]["L8"]["Cloud"]
+    cloud_shadow = em.pixel_flags["pixel_qa"]["L8"]["Cloud Shadow"]
+    
+    # Add up all the mask values
+    all_masked_values = cloud_shadow + cloud + high_cloud_confidence
+    
+    # Apply the masking to the data xarray
+    data_cld_free_xr = data_xr.where(~qa_xr.isin(all_masked_values))
+
+Function 5 - Calculate NDVI and take the mean
+
+    ndvi_xr = es.normalized_diff(data_xr[3], data_xr[2])
+    # IMPORTANT: Replace dataArray NAN with numpy NAN. 
+    # Otherwise there are warnings when creating a dataFrame
+
+# ------------------------------------------------------------ #
+Main code to process mean NDVI time series from multiple sites
+
+# Get the site name from the base folders
+base_path = os.path.join('ndvi-automation',
+                         'sites')
+sitenames = os.listdir(base_path)
```

Comment:
@jacquiewitte This is a great way to avoid the "/" and" \" issues between OS'! I'll have to keep this in mind!